AC-3478 conf-api forms docs (#1678)
livechat/livechat-public-docs: src/pages/management/configuration-api/v3.7/index.mdx

</Section>

# Forms

<Section>

Forms let you read and manage forms configured for a group. Two families of forms are supported:

- **Survey forms** – the pre-chat and post-chat surveys (`prechat`, `postchat`).

> - **Custom forms** – the other chat-widget forms (`offline`, `queued`, `email`, `ask_for_email`, `ask_for_email_mm`).            |

Reviewer: ```suggestion
- **Custom forms** – the other chat widget forms (`offline`, `queued`, `email`, `ask_for_email`, `ask_for_email_mm`).            |
```

Reviewer: here as well ☝️

#### Form

| Parameter | Data type      | Notes                                                                          |
| --------- | -------------- | ------------------------------------------------------------------------------ |
| `id`      | `string`       | The form ID.                                                                   |
| `type`    | `string`       | The form type. One of: `prechat`, `postchat`, `offline`, `queued`, `email`, `ask_for_email`, `ask_for_email_mm`. |
| `enabled` | `bool`         | `true` when the form is enabled. Applies to survey forms only; custom forms don't use it. |

> | `fields`  | `array[Field]` | An array of the form fields, in display order.                                 |

Reviewer: ```suggestion
| `fields`  | `array[Field]` | An array of the form [fields](#####field) in display order.                                 |
```

> ##### Field

Reviewer: ```suggestion
##### Field

A form can contain up to 20 fields. Each radio, select, checkbox, or group field can contain up to 350 answers.
```

| Parameter              | Data type       | Notes                                                                                        |
| ---------------------- | --------------- | -------------------------------------------------------------------------------------------- |

> | `id`                   | `string`        | The field ID. Assigned by the server; ignored when updating.                                 |

Reviewer: ```suggestion
| `id`                   | `string`        | The field ID. Assigned by the server and ignored when updating.                                 |
```

Reviewer: this one left ☝️

> | `type`                 | `string`        | The field type. The allowed values depend on the form `type` (see below).                    |

Reviewer: ```suggestion
| `type`                 | `string`        | The [field type](######field-types). The allowed values depend on the form `type`.                     |
```

Reviewer: and this here ☝️

| `label`                | `string`        | The field label. Required. Must not contain tab characters.                                  |
| `required`             | `bool`          | `true` when the field is required.                                                           |

> | `answers`              | `array[Answer]` | The available answers. Required for `radio`, `select`, `checkbox` and `group` fields.        |

Reviewer: ```suggestion
| `answers`              | `array[Answer]` | The available [answers](######answer). Required for `radio`, `select`, `checkbox`, and `group` fields.        |
```

| `comment_label`        | `string`        | The comment label. Required for `rating` fields. Survey forms only.                          |
| `confirmation_text`    | `string`        | The confirmation text. Required for `email_with_confirmation` fields. Survey forms only.     |
| `confirmation_checked` | `bool`          | Whether the confirmation is checked by default. Applies to `email_with_confirmation` fields. Survey forms only. |

> A form can contain up to 20 fields, and each `radio`, `select`, `checkbox` or `group` field up to 350 answers. Only one `group` field and one `email_with_confirmation` field are allowed per form (survey forms). The `ask_for_email` form must contain an `email` field.

Reviewer: ```suggestion
```

> ###### Field types

Reviewer: ```suggestion
###### Field types

Only one `group` field and one `email_with_confirmation` field are allowed per survey form. The `ask_for_email` form type must contain an `email` field.
```

| `postchat`         | `header`, `question`, `radio`, `select`, `checkbox`, `textarea`, `rating`                       |
| `offline`          | `header`, `name`, `question`, `email`, `radio`, `select`, `checkbox`, `textarea`, `subject`     |
| `ask_for_email`    | `header`, `name`, `question`, `email`, `radio`, `select`, `checkbox`, `textarea`, `subject`     |
| `ask_for_email_mm` | `header`, `name`, `question`, `email`, `radio`, `select`, `checkbox`, `textarea`, `subject`     |
| `queued`           | `header`                                                                                        |
| `email`            | `email`                                                                                         |

###### Answer

| Parameter           | Data type | Notes                                                                                                                                                       |
| ------------------- | --------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------- |

> | `label`             | `string`  | The answer label. Required. Must not contain tab characters.                                                                                                |

Reviewer: ```suggestion
| `label`             | `string`  | The answer label. Required. Can't contain tab characters.                                                                                                |
```

> | `checked`           | `bool`    | Whether the answer is checked by default. Survey forms only: required for `radio`, `select` and `checkbox` fields; for `radio` and `select` at most one answer may be checked. |

Reviewer: ```suggestion
| `checked`           | `bool`    | Whether the answer is checked by default. Survey forms only: required for `radio`, `select`, and `checkbox` fields; for `radio` and `select`, only one answer may be checked. |
```

| -------------------------------- | ---------------------------------------------------------------- |
| Method URL                       | `https://api.livechatinc.com/v3.7/configuration/action/get_form` |
| Required scopes                  | `surveys_read`<sup>**1**</sup> or `forms_manage`<sup>**1**</sup> |
| [Batch support](#batch-requests) | No                                                               |

**1)** The scope depends on the form `type`. Survey forms (`prechat`, `postchat`) require `surveys_read`; custom forms (`offline`, `queued`, `email`, `ask_for_email`, `ask_for_email_mm`) require `forms_manage`.

#### Request

| Parameter  | Required | Data type | Notes                                                |
| ---------- | -------- | --------- | ---------------------------------------------------- |
| `type`     | Yes      | `string`  | The form type. One of: `prechat`, `postchat`, `offline`, `queued`, `email`, `ask_for_email`, `ask_for_email_mm`. |
| `group_id` | No       | `number`  | The ID of the group the form belongs to. Default: 0. |

#### Response

| Parameter | Data type | Notes               |
| --------- | --------- | ------------------- |

> | `form`    | `Form`    | The requested form. |

Reviewer: ```suggestion
| `form`    | `Form`    | The requested [form](####form). |
```

| Required scopes                  | `surveys_manage`<sup>**1**</sup> or `forms_manage`<sup>**1**</sup>  |
| [Batch support](#batch-requests) | No                                                                  |

**1)** The scope depends on the form `type`. Survey forms (`prechat`, `postchat`) require `surveys_manage`; custom forms (`offline`, `queued`, `email`, `ask_for_email`, `ask_for_email_mm`) require `forms_manage`.

#### Request

| Parameter  | Required | Data type      | Notes                                                                               |
| ---------- | -------- | -------------- | ----------------------------------------------------------------------------------- |
| `type`     | Yes      | `string`       | The form type. One of: `prechat`, `postchat`, `offline`, `queued`, `email`, `ask_for_email`, `ask_for_email_mm`. |
| `group_id` | No       | `number`       | The ID of the group the form belongs to. Default: 0.                                |
| `enabled`  | No       | `bool`         | If set to `true`: the form will be enabled. If set to `false`: it will be disabled. Survey forms only; ignored for custom forms. |

> | `fields`   | See notes | `array[Field]` | The form fields. For survey forms it's optional — when omitted, the existing fields are left unchanged. For custom forms it's required and must contain at least one entry. |

Reviewer: ```suggestion
| `fields`   | See notes | `array[Field]` | The form [fields](#####field). For survey forms, it's optional — when omitted, the existing fields are left unchanged. For custom forms, it's required and must contain at least one entry. |
```

#### Response

| Parameter | Data type | Notes             |
| --------- | --------- | ----------------- |

> | `form`    | `Form`    | The updated form. |

Reviewer: ```suggestion
| `form`    | `Form`    | The updated [form](####form). |
```

</Section>

# Forms

<Section>

Forms let you read and manage forms configured for a group. Two families of forms are supported:

- **Survey forms** – the pre-chat and post-chat surveys (`prechat`, `postchat`).
- **Custom forms** – the other chat-widget forms (`offline`, `queued`, `email`, `ask_for_email`, `ask_for_email_mm`).

The required scope depends on the form family, which is determined by the `type` passed in the payload:

| Form family  | `type` values                                                     | Required scopes                   |
| ------------ | ----------------------------------------------------------------- | --------------------------------- |
| Survey forms | `prechat`, `postchat`                                             | `surveys_read` / `surveys_manage` |
| Custom forms | `offline`, `queued`, `email`, `ask_for_email`, `ask_for_email_mm` | `forms_manage`                    |

> 

Reviewer: ```suggestion
```

| `label`                | `string`        | The field label. Required. Must not contain tab characters.                                  |
| `required`             | `bool`          | `true` when the field is required.                                                           |
| `answers`              | `array[Answer]` | The available answers. Required for `radio`, `select`, `checkbox` and `group` fields.        |
| `comment_label`        | `string`        | The comment label. Required for `rating` fields. Survey forms only.                          |
| `confirmation_text`    | `string`        | The confirmation text. Required for `email_with_confirmation` fields. Survey forms only.     |
| `confirmation_checked` | `bool`          | Whether the confirmation is checked by default. Applies to `email_with_confirmation` fields. Survey forms only. |

A form can contain up to 20 fields, and each `radio`, `select`, `checkbox` or `group` field up to 350 answers. Only one `group` field and one `email_with_confirmation` field are allowed per form (survey forms). The `ask_for_email` form must contain an `email` field.

> 

Reviewer: ```suggestion
###### Field types

| Form type          | Allowed field types                                                                             |
| ------------------ | ----------------------------------------------------------------------------------------------- |
| `prechat`          | `header`, `name`, `email`, `email_with_confirmation`, `question`, `radio`, `select`, `checkbox`, `group` |
| `postchat`         | `header`, `question`, `radio`, `select`, `checkbox`, `textarea`, `rating`                       |
| `offline`          | `header`, `name`, `question`, `email`, `radio`, `select`, `checkbox`, `textarea`, `subject`     |
| `ask_for_email`    | `header`, `name`, `question`, `email`, `radio`, `select`, `checkbox`, `textarea`, `subject`     |
| `ask_for_email_mm` | `header`, `name`, `question`, `email`, `radio`, `select`, `checkbox`, `textarea`, `subject`     |
| `queued`           | `header`                                                                                        |
| `email`            | `email`                                                                                         |

```

###### Answer

| Parameter           | Data type | Notes                                                                                                                                                       |
| ------------------- | --------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `label`             | `string`  | The answer label. Required. Must not contain tab characters.                                                                                                |
| `checked`           | `bool`    | Whether the answer is checked by default. Survey forms only: required for `radio`, `select` and `checkbox` fields; for `radio` and `select` at most one answer may be checked. |
| `destination_group` | `number`  | The group the customer is routed to when the answer is selected. Required for `group` field answers. Survey forms only.                                     |

> For custom forms, answers use only the `label` field; `checked` and `destination_group` are ignored.
> 
> ###### Field types
> 
> | Form type          | Allowed field types                                                                             |
> | ------------------ | ----------------------------------------------------------------------------------------------- |
> | `prechat`          | `header`, `name`, `email`, `email_with_confirmation`, `question`, `radio`, `select`, `checkbox`, `group` |
> | `postchat`         | `header`, `question`, `radio`, `select`, `checkbox`, `textarea`, `rating`                       |
> | `offline`          | `header`, `name`, `question`, `email`, `radio`, `select`, `checkbox`, `textarea`, `subject`     |
> | `ask_for_email`    | `header`, `name`, `question`, `email`, `radio`, `select`, `checkbox`, `textarea`, `subject`     |
> | `ask_for_email_mm` | `header`, `name`, `question`, `email`, `radio`, `select`, `checkbox`, `textarea`, `subject`     |
> | `queued`           | `header`                                                                                        |
> | `email`            | `email`                                                                                         |
> 

Reviewer: ```suggestion
```

</Section>

# Forms

<Section>

Forms let you read and manage forms configured for a group. Two families of forms are supported:

- **Survey forms** – the pre-chat and post-chat surveys (`prechat`, `postchat`).
- **Custom forms** – the other chat-widget forms (`offline`, `queued`, `email`, `ask_for_email`, `ask_for_email_mm`).

The required scope depends on the form family, which is determined by the `type` passed in the payload:

| Form family  | `type` values                                                     | Required scopes                   |
| ------------ | ----------------------------------------------------------------- | --------------------------------- |
| Survey forms | `prechat`, `postchat`                                             | `surveys_read` / `surveys_manage` |
| Custom forms | `offline`, `queued`, `email`, `ask_for_email`, `ask_for_email_mm` | `forms_manage`                    |

<Warning>

For custom forms, the `forms_manage` scope is required for both reading (`get_form`) and writing (`update_form`, `delete_form`) — there is no separate read scope.

> </Warning>

Reviewer: I'd remove this from here; it's explained below directly within the methods, so we're duplicating the same thing — I think it's better to tie it contextually to a method rather than a general method group

###### Answer

| Parameter           | Data type | Notes                                                                                                                                                       |
| ------------------- | --------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `label`             | `string`  | The answer label. Required. Must not contain tab characters.                                                                                                |
| `checked`           | `bool`    | Whether the answer is checked by default. Survey forms only: required for `radio`, `select` and `checkbox` fields; for `radio` and `select` at most one answer may be checked. |
| `destination_group` | `number`  | The group the customer is routed to when the answer is selected. Required for `group` field answers. Survey forms only.                                     |

For custom forms, answers use only the `label` field; `checked` and `destination_group` are ignored.

> Allowed field types per form type:

Reviewer: ```suggestion
###### Field types

Allowed field types per form type:
```

| Form type          | Allowed field types                                                                             |
| ------------------ | ----------------------------------------------------------------------------------------------- |
| `prechat`          | `header`, `name`, `email`, `email_with_confirmation`, `question`, `radio`, `select`, `checkbox`, `group` |
| `postchat`         | `header`, `question`, `radio`, `select`, `checkbox`, `textarea`, `rating`                       |
| `offline`          | `header`, `name`, `question`, `email`, `radio`, `select`, `checkbox`, `textarea`, `subject`     |
| `ask_for_email`    | `header`, `name`, `question`, `email`, `radio`, `select`, `checkbox`, `textarea`, `subject`     |
| `ask_for_email_mm` | `header`, `name`, `question`, `email`, `radio`, `select`, `checkbox`, `textarea`, `subject`     |
| `queued`           | `header`                                                                                        |
| `email`            | `email`                                                                                         |

> A form can contain up to 20 fields, and each `radio`, `select`, `checkbox` or `group` field up to 350 answers. Only one `group` field and one `email_with_confirmation` field are allowed per form (survey forms). The `ask_for_email` form must contain an `email` field.

Reviewer: I would move this above the ####Answer header, since looking at the response, the structure is something like survey -> survey fields -> types -> answer so it's more intuitive to put the table order this way

### Get Form

Returns a single form together with its fields.

#### Specifics

|                                  |                                                                  |
| -------------------------------- | ---------------------------------------------------------------- |
| Method URL                       | `https://api.livechatinc.com/v3.7/configuration/action/get_form` |

> | Required scopes                  | `surveys_read`<sup>**1**</sup> `forms_manage`<sup>**1**</sup>    |

Reviewer: ```suggestion
| Required scopes                  | `surveys_read`<sup>**1**</sup> or `forms_manage`<sup>**1**</sup>    |
```

{
  "form": {
    "id": "1",
    "type": "prechat",
    "enabled": true,
    "fields": [
      {
        "id": "153925275950librarydetails",
        "type": "name",
        "label": "Name",
        "required": true
      },
      {
        "id": "153925275950libraryemail",
        "type": "email",
        "label": "E-mail",
        "required": true
      },
      {
        "id": "153925275950librarytopics",
        "type": "checkbox",
        "label": "Pick topics",
        "required": false,
        "answers": [
          { "label": "Billing", "checked": true },
          { "label": "Support", "checked": false }
        ]
      }
    ]
  }
}
```

</CodeResponse>

</Code>

### Update Form

Creates or updates a form. The whole form is replaced with the provided configuration, so include all the fields that should remain. Returns the updated form.

#### Specifics

|                                  |                                                                     |
| -------------------------------- | ------------------------------------------------------------------- |
| Method URL                       | `https://api.livechatinc.com/v3.7/configuration/action/update_form` |

> | Required scopes                  | `surveys_manage`<sup>**1**</sup> `forms_manage`<sup>**1**</sup>     |

Reviewer: ```suggestion
| Required scopes                  | `surveys_manage`<sup>**1**</sup> or `forms_manage`<sup>**1**</sup>     |
```

  -H 'Authorization: Basic <your_personal_access_token>' \
  -H 'Content-Type: application/json' \
  -d '{
    "type": "prechat",
    "group_id": 0,
    "enabled": true,
    "fields": [
      {
        "type": "question",
        "label": "What is your name?",
        "required": true
      },
      {
        "type": "checkbox",
        "label": "Pick topics",
        "required": false,
        "answers": [
          { "label": "Billing", "checked": true },
          { "label": "Support", "checked": false }
        ]
      }
    ]
  }'
```

</CodeSample>

<CodeResponse>

```json
{
  "form": {
    "id": "1",
    "type": "prechat",
    "enabled": true,
    "fields": [
      {
        "id": "153925275950libraryquestion",
        "type": "question",
        "label": "What is your name?",
        "required": true
      },
      {
        "id": "153925275950librarytopics",
        "type": "checkbox",
        "label": "Pick topics",
        "required": false,
        "answers": [
          { "label": "Billing", "checked": true },
          { "label": "Support", "checked": false }
        ]
      }
    ]
  }
}
```

</CodeResponse>

</Code>

> The example below updates a custom form (`ask_for_email`). Custom forms require the `forms_manage` scope, `fields` must contain at least one entry, and answers use only the `label` field.

Reviewer: Personally, I'd keep one example only — wdyt? Which one do you think is better?

Author: will leave the first example only in this case

<CodeResponse>

```json
{
  "form": {
    "id": "0",
    "type": "ask_for_email",
    "enabled": false,
    "fields": [
      {
        "id": "153925275950libraryheader",
        "type": "header",
        "label": "Leave us your e-mail",
        "required": false
      },
      {
        "id": "153925275950libraryemail",
        "type": "email",
        "label": "E-mail",
        "required": true
      },
      {
        "id": "153925275950librarytopics",
        "type": "checkbox",
        "label": "Pick topics",
        "required": false,
        "answers": [
          { "label": "Billing" },
          { "label": "Support" }
        ]
      }
    ]
  }
}
```

</CodeResponse>

</Code>

### Delete Form

Deletes an existing form. For survey forms (`prechat`, `postchat`) the form is deactivated; for custom forms the form configuration is removed.

#### Specifics

|                                  |                                                                     |
| -------------------------------- | ------------------------------------------------------------------- |
| Method URL                       | `https://api.livechatinc.com/v3.7/configuration/action/delete_form` |

> | Required scopes                  | `surveys_manage`<sup>**1**</sup> `forms_manage`<sup>**1**</sup>     |

Reviewer: ```suggestion
| Required scopes                  | `surveys_manage`<sup>**1**</sup> or `forms_manage`<sup>**1**</sup>     |
```
